core/unix/os.c

```diff
@@ -8584,6 +8636,34 @@ post_system_call(dcontext_t *dcontext)
         break;
     }
 #    endif
+#    ifdef LINUX
+    case SYS_ppoll: {
+        kernel_sigset_t app_sigblocked =
+            handle_post_extended_syscall_sigmasks(dcontext, success);
+        /* restore the old syscall parameter's value, which is the emulated sigmask
+         * from DR. */
+        set_syscall_param(dcontext, 3, *(reg_t *)&app_sigblocked);
```

Comment:
This doesn't look right: the parameter should be a pointer to a kernel_sigset_t, not the value of a kernel_sigset_t, and in particular the *original* app pointer value.  It should be saved in pre-syscall.  You can use dcontext->sys_paramN.

If possible, please add a test case to the test to cover this path so we know it works (may require direct call to syscall using syscall() or sthg).

Ditto 2x below.

Reply:
ppoll is part of the test that was added.

Question here, in handle_pre_extended_syscall_sigmasks, we move the app's mask to DR's emulation:

info->pre_syscall_app_sigprocmask = info->app_sigblocked;
signal_set_mask(dcontext, mask);

Can I restore the original emulate mask in DR w/ signal_set_mask(dcontext, info->pre_syscall_app_sigprocmask)? i.e. is info->app_sigblocked the current emulated mask value at a given time?

Reply:
> ppoll is part of the test that was added.

The test does not include a test for examining the parameter post-syscall and thus does not test this code path.  It may need inline asm to do so.

Yes, I did not see any issue with the mask emulation code by inspection, but best to have the test check that also to be sure (call sigprocmask and query the current mask in the test before and after the syscall).

Reply:
So above is ok?
I'll add this to the test then

Reply:
Ok I fixed restoring the original 'mask' pointer into the syscall parameter for all 3 calls and added this to the test as inline asm for X64.